Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DownloadWindow.h"
+#include <AK/Debug.h>
+#include <FileSystemAccessServer/ClientConnection.h>
+#include <LibC/fcntl.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileResponse ClientConnection::request_file(String const& sender, String const& path, int requested_access)
+{
+    // TODO: This hangs the calling application
+
+    // Must receive an absolute path
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    if (m_approved_files.get(path).has_value()) {
+        approved = (requested_access & m_approved_files.get(path).value()) == requested_access;
+    }
+
+    if (!approved) {
+        StringBuilder builder;
+        if (requested_access & O_RDONLY)
+            builder.append('r');
+
+        if (requested_access & O_WRONLY)
+            builder.append('w');
+
+        if (requested_access & O_CREAT)
+            builder.append('c');
+
+        auto access_string = builder.to_string();
+
+        auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            // Intentionally overwrite existing access
+            m_approved_files.set(path, requested_access);
+        }
```

Comment:
What if one part of a process always requests "r" access, while another always requests "w" access. Maybe you could do this, to "merge" the two requested permissions:
```suggestion
        if (approved) {
            auto new_permissions = requested_access;
            if (maybe_permissions.has_value())
                new_permissions |= *maybe_permissions;
            m_approved_files.set(path, new_permissions);
        }
```

Reply:
I'm worried about `O_TRUNC`, `O_APPEND`, `O_EXCL`, and possibly more. Perhaps I should persist everything except for these flags.

Reply:
I'm not worried anymore, the open mode is already controlled by whatever requested access is.